api/lib/opentelemetry/adapter.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  # The basic interface for Adapter objects.
+  #
+  # Adapter class is intended to be subclassed, e.g.,
+  #
+  #   module Adapters
+  #     module Faraday
+  #       class Adapter < OpenTelemetry::Adapter
+  #       end
+  #     end
+  #   end
+  #
+  # then,
+  #
+  #   Adapters::Faraday.install(name: ..., version: ...)
+  class Adapter
+    class << self
+      attr_reader :config,
+                  :http_formatter,
+                  :tracer
+
+      def install(config = {})
+        @config = config
+        @http_formatter = config[:http_formatter] || default_http_formatter
+        @tracer = config[:tracer] || default_tracer
+        new.install
+      end
+
+      private
+
+      def default_http_formatter
```

Comment:
I'm not sure how I feel about this being on the Adapter base class. While many adapters will need to propagate context, not all will. Ideally, formatters would be available on the tracer, and could be accessed when needed through the tracer reference. At one point, they were there, but named tracers caused them to move to the TracerFactory. 

In principle, I like the idea of instrumentation being bound to a tracer, and it having access to everything it needs through it. However, given the current architecture, it might make sense for instrumentation to have direct access to the TracerFactory and just lookup the tracer there. The same considerations apply to meters, when we get around to them.

If the current named tracers / named meters architecture holds, I don't think this Adapter, as its currently implemented is what we need. However, this could be useful if we change directions with named tracers / meters, which does not appear to be out of the question at this time.

Reply:
It would be ideal if we could bind instrumentation to a tracer and meter at install time. If we could do that, I could see this adapter class being the right place to hold that information and the adapter class would have a reference to the config, the tracer, and meter. This would be ideal as it would keep named tracers and named meters out instrumentation.

The thing that I don't like is having to bind to the propagator here, and we should discuss if there are ways around that. One potential way out would be to have an accessor on the Tracer that just delegates to the TracerFactory.

Overall, I would prefer this approach to having to directly use the named tracers / meters in instrumentation. I think this will be a good discussion for tomorrow's SIG meeting. So, to revise my stance, this could be on the right track pending the outcome of our discussion.

Reply:
Continuing to thing out loud, we are currently in the midst of a context propagation overhaul, which is going to lift context propagation to the top level (and off of the TracerFactory), so that actually solves one of the problems. When that work is done, we'll have a Propagation object that will have a reference to an HTTPInjector / Extractor, BinaryInjector / Extractor, and likely a RackInjector / Extractor.

So, I think this is probably going to work for us and its shape will probably look something like:

```ruby
  class Adapter
    class << self
      attr_reader :tracer, :meter, :propagation, :config

      def install(tracer:, meter:, propagation:, config:)
        ...
      end
    end
  end
```

(which is pretty close to what we have)...

Ultimately, I could see the `Adapter` class taking the role of the `Instrumentation` class that I describe in #19. It could override the `inherited` hook to populate an `AdapterRegistry` that could be used to auto-install instrumentation. I could see the `Adapter` class gain a `library_present?` method (for whether or not the underlying gem is present in the app) and an `enabled?` method, for whether or not the instrumentation should be installed.

I think I have a better picture of how these different pieces fit together. We should still discuss whether that picture is accurate or not. If it is, I think we have a reasonable path forward.

Reply:
I've made changes, and am hoping "`propagation_format" is ok, or did you prefer "`propagation`"? :heavy_check_mark:

Reply:
> have an accessor on the Tracer that just delegates to the TracerFactory

I think this is generally useful. Most of the time we will have a Tracer handy, and it'd be convenient to access all the things through it instead of the TracerFactory.